src/timesignature.ts

```diff
@@ -0,0 +1,137 @@
+    );
+  }
+
+  numbers.forEach((number) => {
+    if (isNaN(Number(number))) {
+      throw new Vex.RERR('BadTimeSignature', `Invalid time spec: ${timeSpec}. Must contain two valid numbers.`);
+    }
+  });
+};
+
+export class TimeSignature extends StaveModifier {
+  point?: number;
+  bottomLine?: number;
+  protected timeSig?: TimeSignatureInfo;
+
+  protected validate_args: boolean;
+
+  topLine?: number;
+
+  static get CATEGORY(): string {
+    return 'timesignatures';
+  }
+
+  static get glyphs(): Record<string, { code: string; point: number; line: number }> {
+    return {
+      C: {
+        code: 'timeSigCommon',
+        point: 40,
+        line: 2,
+      },
+      'C|': {
+        code: 'timeSigCutCommon',
+        point: 40,
+        line: 2,
+      },
+    };
+  }
+
+  constructor(timeSpec?: string, customPadding = 15, validate_args = true) {
+    super();
+    this.setAttribute('type', 'TimeSignature');
+    this.validate_args = validate_args;
+
+    if (!timeSpec) return;
+    const padding = customPadding;
+
+    this.point = this.musicFont.lookupMetric('digits.point');
+    const fontLineShift = this.musicFont.lookupMetric('digits.shiftLine', 0);
+    this.topLine = 2 + fontLineShift;
+    this.bottomLine = 4 + fontLineShift;
+    this.setPosition(StaveModifier.Position.BEGIN);
+    this.timeSig = this.parseTimeSpec(timeSpec);
+    this.setWidth(this.getTimeSig().glyph.getMetrics().width!);
+    this.setPadding(padding);
+  }
+
+  getCategory(): string {
+    return TimeSignature.CATEGORY;
+  }
+
+  parseTimeSpec(timeSpec: string): TimeSignatureInfo {
+    if (timeSpec === 'C' || timeSpec === 'C|') {
+      const { line, code, point } = TimeSignature.glyphs[timeSpec];
+      return {
+        line,
+        num: false,
+        glyph: new Glyph(code, point),
+      };
+    }
+
+    if (this.validate_args) {
+      assertIsValidFraction(timeSpec);
+    }
+
+    const [topDigits, botDigits] = timeSpec.split('/').map((number) => number.split(''));
+
+    return {
+      num: true,
+      glyph: this.makeTimeSignatureGlyph(topDigits, botDigits),
+    };
+  }
+
+  makeTimeSignatureGlyph(topDigits: string[], botDigits: string[]): Glyph {
+    const glyph = new TimeSignatureGlyph(this, topDigits, botDigits, 'timeSig0', this.point!);
+    return glyph;
+  }
+
+  getTimeSig(): TimeSignatureInfo {
+    return this.timeSig!;
+  }
+
+  setTimeSig(timeSpec: string): this {
+    this.timeSig = this.parseTimeSpec(timeSpec);
+    return this;
+  }
+
+  draw(): void {
+    if (!this.x) {
+      throw new Vex.RERR('TimeSignatureError', "Can't draw time signature without x.");
+    }
+
+    if (!this.stave) {
+      throw new Vex.RERR('TimeSignatureError', "Can't draw time signature without stave.");
+    }
+
+    this.setRendered();
+    this.timeSig!.glyph.setStave(this.stave);
+    this.timeSig!.glyph.setContext(this.stave.getContext());
+    this.placeGlyphOnLine(this.timeSig!.glyph, this.stave, 2);
```

Comment:
Where is the `2` coming from?

Reply:
It is the value by try and change that makes the `C` & `C|` time signatures stay in the middle. The line has no influence on the position but it does have one on the two mentioned before. `2` is the middle line.

Reply:
Okay, thanks, can you add a comment stating 2 == middle line, and it affects only the `C` and `C|`.